compiler/rustc_mir_transform/src/simplify.rs

```diff
@@ -292,6 +297,168 @@ pub(super) fn simplify_duplicate_switch_targets(terminator: &mut Terminator<'_>)
+    else {
+        // We expect the discriminant to have read `&mut self`,
+        // so we expect the place to be a deref. If we didn't, then
+        // it may have been optimized out, so bail gracefully.
+        return;
+    };
+    if !discr_locals.contains(&deref_local) {
+        // The place being read isn't `_1` (self) or a `Derefer`-inserted local.
+        // It may have been optimized out, so bail gracefully.
+        return;
+    }
+    let TerminatorKind::SwitchInt { discr: Operand::Move(place), targets } = &bb0.terminator().kind
+    else {
+        // When panic=abort, we may end up folding away the other variants of the
+        // coroutine, and end up with ths `SwitchInt` getting replaced. In this
+        // case, there's no need to do this optimization, so bail gracefully.
+        return;
+    };
+    if place != discr_place {
+        // Make sure we don't try to match on some other `SwitchInt`; we should be
+        // matching on the discriminant we just read.
+        return;
+    }
+
+    let mut visited = DenseBitSet::new_empty(body.basic_blocks.len());
+    let mut worklist = vec![];
+    let mut visited_variants = DenseBitSet::new_empty(coroutine_layout.variant_fields.len());
+
+    // Insert unresumed (initial), returned, panicked variants.
+    // We treat these as always reachable.
+    visited_variants.insert(VariantIdx::from_usize(0));
+    visited_variants.insert(VariantIdx::from_usize(1));
+    visited_variants.insert(VariantIdx::from_usize(2));
+    worklist.push(targets.target_for_value(0));
+    worklist.push(targets.target_for_value(1));
+    worklist.push(targets.target_for_value(2));
+
+    // Walk all of the reachable variant blocks.
+    while let Some(block) = worklist.pop() {
+        if !visited.insert(block) {
+            continue;
+        }
+
+        let data = &body.basic_blocks[block];
+        for stmt in &data.statements {
+            match &stmt.kind {
+                // If we see a `SetDiscriminant` statement for our coroutine,
+                // mark that variant as reachable and add it to the worklist.
+                StatementKind::SetDiscriminant { place, variant_index } => {
+                    let PlaceRef { local: deref_local, projection: &[PlaceElem::Deref] } =
+                        (**place).as_ref()
+                    else {
+                        continue;
+                    };
+                    if !discr_locals.contains(&deref_local) {
+                        continue;
+                    }
+                    visited_variants.insert(*variant_index);
+                    worklist.push(targets.target_for_value(variant_index.as_u32().into()));
+                }
+                // The derefer may have inserted a local to access the variant.
+                // Make sure we keep track of it here.
+                StatementKind::Assign(box (place, Rvalue::CopyForDeref(deref_place))) => {
```

Comment:
This could be a Rvalue::Operand too.

Reply:
`Rvalue::Operand(RValue::Copy(_))` would mean that GVN has passed by, and I think GVN would've collapsed all of them into this one: https://github.com/rust-lang/rust/pull/135471/files#diff-2b899097018634da368ec3c772523e10def01aef54b950e07c125a49e374a3e3R332

I can probably assert against it I guess :)